### PR TITLE
refactor(): refactor collection repeat

### PR DESCRIPTION
**Pull request by ajoslin**
Ready to put through the paces. Test in test/html/collection-repeat.html

**Features/Performance Improvements:**

- The user doesn't have to provide height/width now. We use getComputedStyle() on a test element to find a default height/width.
- If height/width are given as constants (either a number or a percent), the repeater will be extra efficient.
  * In this case, there are no dynamic dimensions to calculate.
  * This is like swimming in a tub of buttery smooth scrolling.
- If height/width are given as dynamic getters, the repeater will enter a less efficient mode (collection-repeat used to always be in this mode).
  * In dynamic mode, performance is still much improved from before. We lazily calculate dimensions on demand instead of all up front, and we have streamlined the node render process.
  * End result? Much smoother in dynamic mode, too.

Basically, everything is smoother, and the directive is easier to use.

**TODO:**

- [x] Public documentation (mostly copy/paste of old documentation)
- [x] Unit tests
- [ ] codepens
- [x] Add refresh-images attribute, or perhaps make it always refresh images if it's faster (still need to test more, I've noticed that refreshing images may just make things always faster).

### CODE REVIEW

**Comment by zane-insta** [comment]
Hey @ajoslin I am not sure if this is a good spot to mention this or not but I read through this blog post http://blog.ionic.io/collection-repeat-iteration-two/ and some of the wording was not very clear. I was assuming that the Dynamic List Mode could calculate the actual height of the new item on the fly and I did not realize until after doing a bunch of digging that we still need to pass the height in with a function call. Perhaps someone should update the blog with a an example for the four modes so that it is clear for others.
